crates/parser/src/grammar/functions.rs

```diff
@@ -9,14 +9,26 @@ use crate::node::{Node, Span};
 use crate::{Label, ParseFailed, ParseResult, Parser};
 
 /// Parse a function definition. The optional `pub` qualifier must be parsed by
-/// the caller, and passed in.
-///
-/// # Panics
-/// Panics if the next token isn't `fn`.
-pub fn parse_fn_def(par: &mut Parser, pub_qual: Option<Span>) -> ParseResult<Node<Function>> {
-    let def_tok = par.assert(TokenKind::Fn);
+/// the caller, and passed in. Next token must be `unsafe` or `fn`.
+pub fn parse_fn_def(par: &mut Parser, mut pub_qual: Option<Span>) -> ParseResult<Node<Function>> {
+    let unsafe_qual = par.optional(TokenKind::Unsafe).map(|tok| tok.span);
+    if let Some(pub_) = par.optional(TokenKind::Pub) {
+        let unsafe_span =
+            unsafe_qual.expect("caller must verify that next token is `unsafe` or `fn`");
+
+        par.fancy_error(
+            "`pub` visibility modifier must come before `unsafe`",
+            vec![Label::primary(
+                unsafe_span + pub_.span,
+                "use `pub unsafe` here",
```

Comment:
>I've decided to enforce the rule that contract functions can be either pub or unsafe, but not both

As I understand `pub unsafe` is a compile time error so the error message of `"use pub unsafe here"` seems to be misleading.

Reply:
Ah, understood :+1: